### PR TITLE
feat: add factory and upgrade method

### PR DESCRIPTION
**Pull request by pscott**
- Add the `Factory` contract and tests
- Add the `upgrade` function on the `Space` contract (and tests)
- Add the `StrategyImpl` to create Strategies with empty params `StrategyImpl::from_address()`
- Add the `setup` file that returns a `Config` struct that will (hopefully) make testing files leaner

### CODE REVIEW

File: starknet/src/utils/types.cairo [hunk 1]
```diff
@@ -120,6 +120,23 @@ struct Strategy {
     params: Array<felt252>,
 }
 
+trait StrategyTrait {
```
Comment (Orland0x) [inline starknet/src/utils/types.cairo]: do you think this should be defined in the test folder somewhere? as the trait would only ever be used in tests.

File: starknet/src/tests/test_upgrade.cairo [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+        let (execution_contract_address, _) = deploy_syscall(
+            ExecutorExecutionStrategy::TEST_CLASS_HASH.try_into().unwrap(),
+            0,
+            ArrayTrait::<felt252>::new().span(),
+            false
+        )
+            .unwrap();
+
+        // Set the owner to be the space itself
```
Comment (Orland0x) [inline starknet/src/tests/test_upgrade.cairo]: Set the owner to be the execution strategy

File: starknet/src/utils/types.cairo (hunk 1 above)
Comment (pscott) [reply starknet/src/utils/types.cairo]: bbf7ca6a0345444477f6bbc6adb8e6ea63f5f731

File: starknet/src/tests/test_upgrade.cairo (hunk 2 above)
Comment (pscott) [reply starknet/src/tests/test_upgrade.cairo]: 386c9484ad3e51f6a04354e9c523b10805fd7daa

**Review by Orland0x — APPROVED** [review]
LGTM, uTACK

**Pull request merged**